spec/fastsync/vdd/fastsyncL2.tla

```diff
@@ -0,0 +1,528 @@
+    [pState EXCEPT 
+        !.blocksRequested = pState.blocksRequested \union {msg}    
+    ]  
+
+HandleRequest(msg, pState) ==  
+    IF msg = NoMsg 
+    THEN pState
+    ELSE IF msg.type = "statusRequest" 
+         THEN HandleStatusRequest(msg, pState)
+         ELSE HandleBlockRequest(msg, pState)               
+
+CreateStatusResponse(peer, pState, randomHeight) ==  
+    LET m == 
+        IF peer \in CORRECT 
+        THEN [type |-> "statusResponse", peerId |-> peer, height |-> pState.peerHeights[peer]]
+        ELSE [type |-> "statusResponse", peerId |-> peer, height |-> randomHeight] IN
+    LET npState == 
+        [pState EXCEPT 
+            !.statusRequested = pState.statusRequested \ {peer}       
+        ] IN
+    [msg |-> m, peers |-> npState] 
+             
+CreateBlockResponse(msg, pState, randomBlock) == 
+    LET m == 
+        IF msg.peerId \in CORRECT
+        THEN [type |-> "blockResponse", peerId |-> msg.peerId, block |-> CorrectBlock(msg.height)]
+        ELSE [type |-> "blockResponse", peerId |-> msg.peerId, block |-> randomBlock] IN
+    LET npState == 
+        [pState EXCEPT 
+            !.blocksRequested = pState.blocksRequested \ {msg}      
+        ] IN
+    [msg |-> m, peers |-> npState]   
+
+CreateResponse(pState) == 
+    \/ /\ pState.statusRequested /= {}  
+       /\ pState.blocksRequested /= {}
+       /\ \E t \in TurnResponse:
+            \/ /\ t = "status"
+               /\ \E h \in Heights: 
+                  \E peer \in pState.statusRequested:   
+                    LET msgAndPeers == CreateStatusResponse(peer, pState, h) IN
+                    /\ peersState' = msgAndPeers.peers
+                    /\ inMsg' = msgAndPeers.msg   
+                                                               
+            \/ /\ t = "block"        
+               /\ \E msg \in pState.blocksRequested:
+                   \E block \in Blocks:
+                    LET msgAndPeers == CreateBlockResponse(msg, pState, block) IN      
+                    /\ peersState' = msgAndPeers.peers
+                    /\ inMsg' = msgAndPeers.msg  
+    
+    \/ /\ pState.statusRequested /= {}  
+       /\ pState.blocksRequested = {}
+       /\ \E h \in Heights: 
+            \E peer \in pState.statusRequested:   
+                LET msgAndPeers == CreateStatusResponse(peer, pState, h) IN
+                /\ peersState' = msgAndPeers.peers
+                /\ inMsg' = msgAndPeers.msg     
+    
+    
+    \/ /\ pState.statusRequested = {} 
+       /\ pState.blocksRequested /= {}
+       /\ \E msg \in pState.blocksRequested:
+            \E block \in Blocks:
+                LET msgAndPeers == CreateBlockResponse(msg, pState, block) IN      
+                /\ peersState' = msgAndPeers.peers
+                /\ inMsg' = msgAndPeers.msg  
+                    
+    \/ /\ pState.statusRequested = {}
+       /\ pState.blocksRequested = {}
+       /\ peersState' = pState
+       /\ inMsg' = NoMsg                  
```

Comment:
This is very nice. Essentially, it is a non-deterministic scheduler for peer responses. Shall it be reflected in the name?